src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_huhuran.cpp

```diff
@@ -15,148 +15,144 @@
         {
-            DoCastVictim(SPELL_ACIDSPIT);
-            Spit_Timer = urand(5000, 10000);
+            DoCastSelf(SPELL_BERSERK, true);
+            Talk(EMOTE_BERSERK);
+            berserk = true;
         }
-        else Spit_Timer -= diff;
+    }
 
-        //NoxiousPoison_Timer
-        if (NoxiousPoison_Timer <= diff)
-        {
-            DoCastVictim(SPELL_NOXIOUSPOISON);
-            NoxiousPoison_Timer = urand(12000, 24000);
-        }
-        else NoxiousPoison_Timer -= diff;
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim())
+            return;
 
-        //PoisonBolt only if frenzy or berserk
-        if (Frenzy || Berserk)
+        events.Update(diff);
+        while (uint32 eventid = events.ExecuteEvent())
         {
-            if (PoisonBolt_Timer <= diff)
+            switch (eventid)
             {
-                DoCastVictim(SPELL_POISONBOLT);
-                PoisonBolt_Timer = 3000;
+                case EVENT_FRENZY:
+                    DoCastSelf(SPELL_FRENZY, true);
+                    Talk(EMOTE_FRENZY_KILL);
+                    events.RepeatEvent(urand(25000, 35000));
+                    break;
+                case EVENT_WYVERN_STING:
+                    me->CastCustomSpell(SPELL_WYVERN_STING, SPELLVALUE_MAX_TARGETS, 10, me, true);
+                    events.RepeatEvent(urand(15000, 32000));
+                    break;
+                case EVENT_ACID_SPIT:
+                    DoCastVictim(SPELL_ACID_SPIT);
+                    events.RepeatEvent(urand(5000, 10000));
+                    break;
+                case EVENT_NOXIOUS_POISON:
+                    if (Unit* target = SelectTarget(SelectTargetMethod::Random, 0, 100, true))
+                    {
+                        DoCast(target, SPELL_NOXIOUS_POISON);
+                    }
+                    events.RepeatEvent(urand(12000, 24000));
+                    break;
+                case EVENT_HARD_ENRAGE:
+                    if (!hardEnrage)
+                    {
+                        DoCastSelf(SPELL_HARD_ENRAGE, true);
+                        hardEnrage = true;
+                    }
+                    else
+                    {
+                        DoCastAOE(SPELL_POISON_BOLT);
+                    }
+                    events.RepeatEvent(3000);
+                    break;
             }
-            else PoisonBolt_Timer -= diff;
-        }
-
-        //FrenzyBack_Timer
-        if (Frenzy && FrenzyBack_Timer <= diff)
-        {
-            me->InterruptNonMeleeSpells(false);
-            Frenzy = false;
-            FrenzyBack_Timer = 15000;
         }
-        else FrenzyBack_Timer -= diff;
-
-        if (!Berserk && HealthBelowPct(31))
-        {
-            me->InterruptNonMeleeSpells(false);
-            Talk(EMOTE_BERSERK);
-            DoCast(me, SPELL_BERSERK);
-            Berserk = true;
-        }
-
         DoMeleeAttackIfReady();
     }
+    private:
```

Comment:
I'd rather use protected instead of private, but that's more of a personal choice (in this case).

Reply:
skip then.